internal/refactoring/move_validate.go

```diff
@@ -200,30 +205,55 @@ func ValidateMoves(stmts []MoveStatement, rootCfg *configs.Config, declaredInsts
 	if !diags.HasErrors() {
-		for _, cycle := range g.Cycles() {
-			// Reporting cycles is awkward because there isn't any definitive
-			// way to decide which of the objects in the cycle is the cause of
-			// the problem. Therefore we'll just list them all out and leave
-			// the user to figure it out. :(
-			stmtStrs := make([]string, 0, len(cycle))
-			for _, stmtI := range cycle {
-				// move statement graph nodes are pointers to move statements
-				stmt := stmtI.(*MoveStatement)
-				stmtStrs = append(stmtStrs, fmt.Sprintf(
-					"\n  - %s: %s → %s",
-					stmt.DeclRange.StartString(),
-					stmt.From.String(),
-					stmt.To.String(),
-				))
-			}
-			sort.Strings(stmtStrs) // just to make the order deterministic
+		diags = diags.Append(validateMoveStatementGraph(g))
+	}
+
+	return diags
+}
+
+func validateMoveStatementGraph(g *dag.AcyclicGraph) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+	for _, cycle := range g.Cycles() {
+		// Reporting cycles is awkward because there isn't any definitive
+		// way to decide which of the objects in the cycle is the cause of
+		// the problem. Therefore we'll just list them all out and leave
+		// the user to figure it out. :(
+		stmtStrs := make([]string, 0, len(cycle))
+		for _, stmtI := range cycle {
+			// move statement graph nodes are pointers to move statements
+			stmt := stmtI.(*MoveStatement)
+			stmtStrs = append(stmtStrs, fmt.Sprintf(
+				"\n  - %s: %s → %s",
+				stmt.DeclRange.StartString(),
+				stmt.From.String(),
+				stmt.To.String(),
+			))
+		}
+		sort.Strings(stmtStrs) // just to make the order deterministic
+
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Cyclic dependency in move statements",
+			fmt.Sprintf(
+				"The following chained move statements form a cycle, and so there is no final location to move objects to:%s\n\nA chain of move statements must end with an address that doesn't appear in any other statements, and which typically also refers to an object still declared in the configuration.",
+				strings.Join(stmtStrs, ""),
+			),
+		))
+	}
 
+	// Look for cycles to self.
+	// A user shouldn't be able to create self-references, but we cannot
+	// correctly process a graph with them.
+	for _, e := range g.Edges() {
+		src := e.Source()
+		if src == e.Target() {
 			diags = diags.Append(tfdiags.Sourceless(
 				tfdiags.Error,
-				"Cyclic dependency in move statements",
+				"Self reference in move statements",
 				fmt.Sprintf(
-					"The following chained move statements form a cycle, and so there is no final location to move objects to:%s\n\nA chain of move statements must end with an address that doesn't appear in any other statements, and which typically also refers to an object still declared in the configuration.",
-					strings.Join(stmtStrs, ""),
+					"The move statement [%s] is referencing itself, and cannot be processed\n"+
+						"This is a bug in Terraform and should be reported",
+					src.(*MoveStatement).Name(),
```

Comment:
If this was intended to be two paragraphs then typically it should have `\n\n` between them to create the usual blank line delimiter. Otherwise, I'd suggest to print this all out as one paragraph.

Also, in recent incarnations of this we've started saying it instead as "This is a bug in Terraform; please report it!", and so it might be good to be consistent with that so that we have consistent messaging for situations where we're asking explicitly for a bug report. (IIRC, we started using this style to make it be active instead of passive voice.)

Putting that together, I guess I can suggest a concrete replacement message for ease of updating it, though feel free to adjust as you see fit!

```suggestion
					"The move statement %s refers to itself the move dependency graph, which is invalid. This is a bug in Terraform; please report it!",
					src.(*MoveStatement).Name(),
```

Reply:
Thanks. We should probably fixup all the the old-style error instances together, which is where most recently saw this wording, otherwise it will probably continue to perpetuate itself.